xmake/rules/c++/modules/modules_support/clang.lua

```diff
@@ -0,0 +1,384 @@
+                table.join2(args, {  "-I", path.directory(headerunit.path), "-x", "c++-user-header", headerunit.path })
+            elseif headerunit.type == ":angle" then
+                table.join2(args, { "-x", "c++-system-header", headerunit.name })
+            end
+
+            batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
+            batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args))
+
+            batchcmds:add_depfiles(headerunit.path)
+            batchcmds:set_depmtime(os.mtime(bmifile))
+            batchcmds:set_depcache(target:dependfile(bmifile))
+
+            table.append(public_flags, modulefileflag .. bmifile)
+        else
+            local bmifile = path.join(stlcachedir, headerunit.name .. get_bmi_ext())
+
+            if not os.isfile(bmifile) then
+                local args = { modulecachepathflag .. stlcachedir, "-c", "-o", bmifile, "-x", "c++-system-header", headerunit.path }
+
+                batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
+                batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args))
+
+            end
+
+            batchcmds:set_depmtime(os.mtime(bmifile))
+            batchcmds:set_depcache(target:dependfile(bmifile))
+
+            table.append(private_flags, modulefileflag .. bmifile)
+        end
+    end
+    return public_flags, private_flags
+end
+
+-- build module files
+function build_modules(target, batchcmds, objectfiles, modules, opt)
+    local compinst = target:compiler("cxx")
+    local cachedir = common.get_cache_dir(target)
+
+    -- get modules flags
+    local modulecachepathflag = get_modulecachepathflag(target)
+    local emitmoduleinterfaceflag = get_emitmoduleinterfaceflag(target)
+    local modulefileflag = get_modulefileflag(target)
+
+    -- append deps modules
+    local flags = {}
+    for _, dep in ipairs(target:orderdeps()) do
+        table.join2(flags, dep:data("cxx.modules.flags"))
+    end
+    flags = table.unique(flags)
+    target:add("cxxflags", flags, {force = true, expand = false})
+
+    local common_args = { modulecachepathflag .. cachedir }
+    for _, objectfile in ipairs(objectfiles) do
+        local m = modules[objectfile]
+
+        if m then
+            if not os.isdir(path.directory(objectfile)) then
+                os.mkdir(path.directory(objectfile))
+            end
+
+            local args = { emitmoduleinterfaceflag }
+            local flag = {}
+            local bmifiles = {}
+            for name, provide in pairs(m.provides) do
+                batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.module.bmi %s", name)
+
+                local bmifile = provide.bmi
+                table.join2(args, { "-c", "-x", "c++-module", "--precompile", provide.sourcefile, "-o", bmifile })
```

Comment:
I don't really understand here, if there are multiple values for provides, then there seems to be some problems.

```
clang -c sourcefile -o bmifile -c sourcefile2 -o bmifile2 ...?
```